src/extensions/idle.rs

```diff
@@ -93,16 +93,26 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
     /// This is necessary so that we can keep using the inner `Session` in `wait_keepalive`.
     fn wait_inner(&mut self) -> Result<()> {
         let mut v = Vec::new();
-        match self.session.readline(&mut v).map(|_| ()) {
-            Err(Error::Io(ref e))
-                if e.kind() == io::ErrorKind::TimedOut || e.kind() == io::ErrorKind::WouldBlock =>
-            {
-                // we need to refresh the IDLE connection
-                self.terminate()?;
-                self.init()?;
-                self.wait_inner()
+        loop {
+            let res = match self.session.readline(&mut v).map(|_| ()) {
+                Err(Error::Io(ref e))
+                    if e.kind() == io::ErrorKind::TimedOut
+                        || e.kind() == io::ErrorKind::WouldBlock =>
+                {
+                    // we need to refresh the IDLE connection
+                    self.terminate()?;
+                    self.init()?;
+                    self.wait_inner()
+                }
+                r => r,
+            };
+
+            // Handle Dovecot's imap_idle_notify_interval message
+            if res.is_ok() && v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
```

Comment:
I think it'd probably be better to use `?` on the `res` above. That way you know that you're in the `Ok` case, and wouldn't need to check for `is_ok()`. And then in the `else` case you'd just `break Ok(())`.

Reply:
Oh yeah, that is better!